GEODE-8877: Configurable UDP membership address

alb3rtobr opened the pull request:
This PR allows to bind membership endpoints to all interfaces via the `--bind-address` attribute. When a wildcard address is introduced with that parameter, the membership UDP traffic will be bound to localhost due to JGroups does not allow to bind to wildcard addresses.

If a different address for UDP traffic needs to be used, it can be specified with the new parameter called `membership-bind-addresss`.

I have not squashed the commits of the PR because I think it can be useful for reviewers. The two first commits contains the functionality I mentioned previously, while the three next commits contains some code improvements identified after the review of [my first PR ](https://github.com/apache/geode/pull/5970)for this ticket.

albertogpz commented on geode-common/src/main/java/org/apache/geode/internal/inet/LocalHostUtil.java:
```diff
@@ -242,6 +242,10 @@ public static InetAddress getAnyLocalAddress() {
   }
 
+  public static boolean isWildcardAddress(String address) {
+    return (address != null && (address.equals("0.0.0.0") || address.equals("::")));
```
How about simplifying this as follows?:
return "0.0.0.0".equals(address) || "::".equals(address);

alb3rtobr replied:
Good catch! I have just added a commit fixing this. Thanks!

albertogpz reviewed (approved):
Looks good to me other than one small comment.

davebarnes97 commented on geode-docs/tools_modules/gfsh/command-pages/start.html.md.erb:
```diff
@@ -526,7 +538,11 @@ start server --name=value [--assign-buckets(=value)] [--bind-address=value]
-<td>The IP address on which the server will be bound.</td>
+<td>The IP address on which the server will be bound. The '*' character is accepted for wildcard addresses.
+<div class="note note">
+<b>Note:</b> In case the server is bound to all local addresses, UDP membership-related traffic will be bound to the local host machine's address.
```
Slight rewrite to use present tense and clarify default:
When the server is bound to all local addresses, the default for UDP membership-related traffic is to bind to the local host machine's address.

davebarnes97 commented on geode-docs/tools_modules/gfsh/command-pages/start.html.md.erb:
```diff
@@ -605,6 +621,14 @@ start server --name=value [--assign-buckets(=value)] [--bind-address=value]
+<td><span class="keyword parmname">\-\-membership-bind-address</span></td>
+<td>Specifies the IP address to which the UDP membership-related traffic will be bound. If defined, it has precedence over the value of <code>bind-address</code>. Wildcard addresses are not accepted.
+<div class="note note">
+<b>Note:</b> By default, UDP membership-related traffic is bound to <code>bind-address</code> but there is one exception. In case the server is bound to all local addresses, UDP membership-related traffic will be bound to the local host machine's address by default.
```
Delete the note here, which is redundant, as it repeats the description of default behavior already explained in the --bind-address entry above.

davebarnes97 reviewed (changes requested):
Please see 2 recommended re-writes in the doc file geode-docs/tools_modules/gfsh/command-pages/start.html.md.erb.

alb3rtobr commented:
> Please see 2 recommended re-writes in the doc file geode-docs/tools_modules/gfsh/command-pages/start.html.md.erb.

Done, thanks Dave!

mhansonp commented on geode-gfsh/src/integrationTest/java/org/apache/geode/management/internal/cli/GfshParserAutoCompletionIntegrationTest.java:
```diff
@@ -30,6 +30,12 @@
+   * Number of @CliOption parameters of StartServerCommand.startServer()
+   * method +1 due to "--group" & "--groups" are defined in the same @CliOption
+   */
+  final int startServerCommandCliOptions = 55;
```
Can this be a calculated rather than magic number?

mhansonp commented on geode-membership/src/main/java/org/apache/geode/distributed/internal/membership/gms/fd/GMSHealthMonitor.java:
```diff
@@ -953,8 +955,22 @@ public void init(Services<ID> s) throws MembershipConfigurationException {
     try {
-      serverSocket = createServerSocket(localAddress.getInetAddress(),
-          services.getConfig().getMembershipPortRange());
+      InetAddress address = localAddress.getInetAddress();
```
I think it would be good to make this a separate function. This clutters up a function that doesn't really care.

mhansonp reviewed (approved):
I made a few comments in this PR none of them are critical. I liked that you cleaned out the use of the Gemfire name and that you got rid of some of the magic numbers by single sourcing them.

boglesby commented on geode-common/src/test/java/org/apache/geode/internal/inet/LocalHostUtilTest.java:
```diff
@@ -0,0 +1,42 @@
+  public void isWildCardAddressReturnsFalseIfAddressIsEmpty() {
+    assertThat(LocalHostUtil.isWildcardAddress("")).isFalse();
+  }
+}
```
Do you need a test like:
```
@Test
public void isWildCardAddressReturnsFalseIfAddressIsNotWildcard() {
  assertThat(LocalHostUtil.isWildcardAddress("1.2.3.4")).isFalse();
}
```

boglesby reviewed (approved):
I only reviewed the two files I own in this PR:

- LocalHostUtil.java
- LocalHostUtilTest.java

I have not reviewed the functionality of it.

alb3rtobr commented:
Thanks @mhansonp & @boglesby , I have added a commit after your comments.

alb3rtobr commented:
@davebarnes97 Could you check if your "request for changes" can be removed? Thanks

mhansonp commented:
> Thanks @mhansonp & @boglesby , I have added a commit after your comments.

I still approve of your changes. Thanks for making the suggested changes as well, they look good.

kirklund reviewed (commented):
I've read the ticket and made a first pass over the diffs, but I don't understand why StartLocatorCommandWorkingDirectoryTest and StartServerCommandWorkingDirectoryTest need to be deleted in order to support configurable UDP membership address?

alb3rtobr commented:
> I've read the ticket and made a first pass over the diffs, but I don't understand why StartLocatorCommandWorkingDirectoryTest and StartServerCommandWorkingDirectoryTest need to be deleted in order to support configurable UDP membership address?

It is not needed. Its a code improvement I identified when implementing the functionality. Notice what I wrote on the PR description:

> I have not squashed the commits of the PR because I think it can be useful for reviewers. The two first commits contains the functionality I mentioned previously, while the three next commits contains some code improvements identified after the review of my first PR for this ticket.

Long story: I observed that adding a new parameter in `start locator` or `start server` commands had an impact on all tests of `StartLocatorCommandWorkingDirectoryTest` and `StartLocatorCommandWorkingDirectoryTest`. Those tests were just testing the working directory resolution, it has no sense they were impacted. So I refactored that code in `StartLocatorCommand` & `StartServerCommand` extracting the working directory resolution code to a new method to be able to test it independently. You can review just what I changed for this in this commit:

`4cd39cc Refactor StartMemberUtils.resolveWorkingDir`

alb3rtobr commented:
I have added an empty commit to retrigger CI because test were executed again today and one test failed unexpectedly. Last commit was done 21 days ago and tests were ok.

onichols-pivotal commented:
> Tests were executed again today and one test failed unexpectedly. Last commit was done 21 days ago

It's always a risk to merge based on tests that passed long time ago (sometimes even short time ago -- a lot can change on develop in a short time).  We recently had a commit merged that immediately broke develop, and found that PR checks had last been run 4 weeks earlier.  To reduce this risk, I will be re-executing tests weekly to improve confidence that PRs have been tested against something more closely resembling the latest develop.

alb3rtobr commented:
> > Tests were executed again today and one test failed unexpectedly. Last commit was done 21 days ago
> 
> It's always a risk to merge based on tests that passed long time ago (sometimes even short time ago -- a lot can change on develop in a short time). We recently had a commit merged that immediately broke develop, and found that PR checks had last been run 4 weeks earlier. To reduce this risk, I will be re-executing tests weekly to improve confidence that PRs have been tested against something more closely resembling the latest develop.

Agree, but then PRs should be rebased against develop branch, right?
In my case, I have seen there are conflicts I have to solve, so Im going to rebase anyway...

onichols-pivotal commented:
>then PRs should be rebased against develop branch, right?

PR checks do automatically rebase against latest develop, so just re-executing them is sufficient.  However when there is a conflict then they will not run at all, so it's a good idea to manually rebase especially if GitHub is also showing conflicts.

alb3rtobr commented:
I have removed two of the code improvements not related with the feature to separated tickets and PRs. I hope this will make the review of this PR easier.
